UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Validation/ConfigValidator.cs

```diff
@@ -0,0 +1,173 @@
+                });
+
+            RuleFor(config => config.AllowedUsers).Must((config, s) => IsAclProvided(config)).WithMessage(ValidationMessageInvalidAcl);
+
+            RuleFor(config => config.AllowedUserGroups).Must((config, s) => IsAclProvided(config)).WithMessage(ValidationMessageInvalidAcl);
+
+            RuleFor(config => config.Tags)
+                .Must(tags => tags != null && tags.Any()).WithMessage(ValidationMessageInvalidOrNullAttribute)
+                .DependentRules(() =>
+                {
+                    RuleFor(config => config.Tags)
+                        .Must(tags => tags.All(tag =>
+                            !string.IsNullOrWhiteSpace(tag?.Key) && !string.IsNullOrWhiteSpace(tag?.Value)))
+                        .WithMessage("Key or Value not provided");
+                });
+
+            RuleFor(config => config.ReadMeSearchFilter)
+                .Must(readMeSearchFilter => !string.IsNullOrEmpty(readMeSearchFilter?.Trim()))
+                .WithMessage(ValidationMessageInvalidOrNullAttribute);
+
+            RuleFor(config => config.BatchExpiryInDays).NotEmpty()
+                .WithMessage(ValidationMessageInvalidOrNullAttribute)
+                .DependentRules(() =>
+                {
+                    RuleFor(config => config.BatchExpiryInDays).GreaterThan(0)
+                        .WithMessage("Expected value is natural number, i.e. number greater than 0");
+                });
+        }
+
+        [ExcludeFromCodeCoverage]
+        private static bool IsAclProvided(BessConfig c)
+        {
+            if ((c.AllowedUsers == null && c.AllowedUserGroups == null) ||
+                    (c.AllowedUsers?.Count() == 0 && c.AllowedUserGroups?.Count() == 0) ||
+                    (c.AllowedUsers == null && c.AllowedUserGroups?.Count() == 0) ||
+                    (c.AllowedUsers?.Count() == 0 && c.AllowedUserGroups == null))
+            {
+                return false;
+            }
+
+            if (c.AllowedUsers == null && c.AllowedUserGroups != null)
+            {
+                if (!c.AllowedUserGroups.Any() || c.AllowedUserGroups.All(group => group?.Trim() is null or "") || c.AllowedUserGroups.Contains(string.Empty) || c.AllowedUserGroups.Contains(null))
+                    return false;
+            }
+
+            if (c.AllowedUserGroups == null && c.AllowedUsers != null)
+            {
+                if (!c.AllowedUsers.Any() || c.AllowedUsers.All(user => user?.Trim() is null or "") || c.AllowedUsers.Contains(string.Empty) || c.AllowedUsers.Contains(null))
+                    return false;
+            }
+
+            if (c.AllowedUserGroups != null && c.AllowedUsers != null)
+            {
+                if (c.AllowedUsers.All(user => user?.Trim() is null or "") && c.AllowedUserGroups.All(group => group?.Trim() is null or ""))
+                    return false;
+            }
```

Comment:
so this too could be written out neater in an if else

Reply:
We had tried this, but due to validation requirements for (AllowedUsers and AllowedUserGroups) we had to move this code to nested if.